File: src-wails/node.go
package main

import (
	"database/sql"
	"fmt"
)

func (a *App) CreateChildNode(projectName string, parentNodeID any, name string) (*NodeInfo, error) {
	var projectID int
	err := db.QueryRow("SELECT id FROM wt_project WHERE name = ?", projectName).Scan(&projectID)
	if err != nil {
		return nil, fmt.Errorf("项目不存在")
	}
	var parentID int
	switch v := parentNodeID.(type) {
	case float64:
		parentID = int(v)
	case int:
		parentID = v
	}
	var parentIDPtr *int
	if parentID > 0 {
		var parentName string
		err = db.QueryRow("SELECT name FROM wt_node WHERE id = ?", parentID).Scan(&parentName)
		if err != nil {
			return nil, fmt.Errorf("父节点不存在")
		}
		parentIDPtr = &parentID
	}
	result, err := db.Exec(`
		INSERT INTO wt_node (project_id, parent_id, name, desc) VALUES (?, ?, ?, '')
	`, projectID, parentIDPtr, name)
	if err != nil {
		return nil, err
	}
	nodeID, _ := result.LastInsertId()
	return &NodeInfo{
		ID:       int(nodeID),
		ParentID: parentIDPtr,
		Name:     name,
	}, nil
}

func (a *App) CreateBrotherNode(projectName string, siblingNodeID any, name string) (*NodeInfo, error) {
	var projectID int
	err := db.QueryRow("SELECT id FROM wt_project WHERE name = ?", projectName).Scan(&projectID)
	if err != nil {
		return nil, fmt.Errorf("项目不存在")
	}
	var siblingID int
	switch v := siblingNodeID.(type) {
	case float64:
		siblingID = int(v)
	case int:
		siblingID = v
	}
	var parentID sql.NullInt64
	err = db.QueryRow("SELECT parent_id FROM wt_node WHERE id = ?", siblingID).Scan(&parentID)
	if err != nil {
		return nil, fmt.Errorf("节点不存在")
	}
	var parentIDVal any
	if parentID.Valid {
		parentIDVal = parentID.Int64
	}
	result, err := db.Exec(`
		INSERT INTO wt_node (project_id, parent_id, name, desc) VALUES (?, ?, ?, '')
	`, projectID, parentIDVal, name)
	if err != nil {
		return nil, err
	}
	nodeID, _ := result.LastInsertId()
	var parentIDPtr *int
	if parentID.Valid {
		pid := int(parentID.Int64)
		parentIDPtr = &pid
	}
	return &NodeInfo{
		ID:       int(nodeID),
		ParentID: parentIDPtr,
		Name:     name,
	}, nil
}

func (a *App) UpdateNodeByID(nodeID int, name string, desc string) error {
	_, err := db.Exec(`UPDATE wt_node SET name = ?, desc = ? WHERE id = ?`, name, desc, nodeID)
	return err
}

func (a *App) UpdateNode(projectName string, nodeData map[string]any) error {
	nodeID, ok := nodeData["id"]
	if !ok {
		return fmt.Errorf("节点数据缺少id")
	}
	var id int
	switch v := nodeID.(type) {
	case float64:
		id = int(v)
	case int:
		id = v
	}
	name, _ := nodeData["name"].(string)
	desc, _ := nodeData["note"].(string)
	_, err := db.Exec(`UPDATE wt_node SET name = ?, desc = ? WHERE id = ?`, name, desc, id)
	if err != nil {
		return err
	}
	return nil
}

func (a *App) DeleteNode(projectName string, nodeID any) error {
	var id int
	switch v := nodeID.(type) {
	case float64:
		id = int(v)
	case int:
		id = v
	case string:
		return fmt.Errorf("invalid node ID type")
	}
	var projectID int
	err := db.QueryRow("SELECT project_id FROM wt_node WHERE id = ?", id).Scan(&projectID)
	if err != nil {
		return err
	}
	var count int
	db.QueryRow("SELECT COUNT(*) FROM wt_node WHERE project_id = ?", projectID).Scan(&count)
	if count <= 1 {
		return fmt.Errorf("不能删除所有节点")
	}
	_, err = db.Exec("DELETE FROM wt_node WHERE id = ?", id)
	return err
}

func (a *App) UpdateNodeNote(projectName string, nodeID any, note string) error {
	var id int
	switch v := nodeID.(type) {
	case float64:
		id = int(v)
	case int:
		id = v
	}
	_, err := db.Exec("UPDATE wt_node SET desc = ? WHERE id = ?", note, id)
	return err
}

func (a *App) RenameNode(nodeID int, newName string) error {
	_, err := db.Exec("UPDATE wt_node SET name = ? WHERE id = ?", newName, nodeID)
	return err
}

func (a *App) UpdateNodeDesc(nodeID int, desc string) error {
	_, err := db.Exec("UPDATE wt_node SET desc = ? WHERE id = ?", desc, nodeID)
	return err
}

func (a *App) RebaseNode(nodeID int, newParentID *int) error {
	if newParentID == nil {
		_, err := db.Exec("UPDATE wt_node SET parent_id = NULL WHERE id = ?", nodeID)
		return err
	}
	current := *newParentID
	for {
		if current == nodeID {
			return fmt.Errorf("不能形成循环引用")
		}
		var parentID sql.NullInt64
		err := db.QueryRow("SELECT parent_id FROM wt_node WHERE id = ?", current).Scan(&parentID)
		if err != nil || !parentID.Valid {
			break
		}
		current = int(parentID.Int64)
	}
	_, err := db.Exec("UPDATE wt_node SET parent_id = ? WHERE id = ?", *newParentID, nodeID)
	return err
}

func (a *App) GetBranchTag(parentID int, childID int) (*WTBranchTag, error) {
	var tag WTBranchTag
	err := db.QueryRow(`
		SELECT id, parent_id, child_id, name, COALESCE(desc, '')
		FROM wt_branch_tag WHERE parent_id = ? AND child_id = ?
	`, parentID, childID).Scan(&tag.ID, &tag.ParentID, &tag.ChildID, &tag.Name, &tag.Desc)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &tag, nil
}

func (a *App) SetBranchTag(parentID int, childID int, name string, desc string) error {
	_, err := db.Exec(`
		INSERT INTO wt_branch_tag (parent_id, child_id, name, desc) VALUES (?, ?, ?, ?)
		ON CONFLICT(parent_id, child_id) DO UPDATE SET name = excluded.name, desc = excluded.desc
	`, parentID, childID, name, desc)
	return err
}

func (a *App) DeleteBranchTag(parentID int, childID int) error {
	_, err := db.Exec("DELETE FROM wt_branch_tag WHERE parent_id = ? AND child_id = ?", parentID, childID)
	return err
}

func (a *App) GetNodePath(nodeID int) ([]int, error) {
	fmt.Println("[GetNodePath] 开始, nodeID:", nodeID)
	var path []int
	current := nodeID
	for {
		fmt.Println("[GetNodePath] 处理current:", current)
		path = append([]int{current}, path...)
		var parentID sql.NullInt64
		fmt.Println("[GetNodePath] 查询parent_id...")
		err := db.QueryRow("SELECT parent_id FROM wt_node WHERE id = ?", current).Scan(&parentID)
		fmt.Println("[GetNodePath] 查询完成, err:", err, "parentID.Valid:", parentID.Valid)
		if err != nil || !parentID.Valid {
			break
		}
		current = int(parentID.Int64)
	}
	fmt.Println("[GetNodePath] 完成, path:", path)
	return path, nil
}

func (a *App) GenerateLineSerial(projectName string) (string, error) {
	var projectID int
	err := db.QueryRow("SELECT id FROM wt_project WHERE name = ?", projectName).Scan(&projectID)
	if err != nil {
		projectID = 0
	}
	tx, err := db.Begin()
	if err != nil {
		return "", err
	}
	defer tx.Rollback()
	sn := generateSerial(tx, projectID)
	return sn, nil
}

File: src-tauri/src/commands/node.rs
use crate::database;
use crate::types::WTNode;
const SERIAL_CHARS: &[u8] = b"abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";

#[tauri::command]
pub fn create_child_node(project_name: String, parent_id: i64, name: String) -> Result<WTNode, String> {
    database::with_db(|conn| {
        let project_id: i64 = conn.query_row(
            "SELECT id FROM wt_project WHERE name = ?",
            [&project_name],
            |row| row.get(0),
        )?;
        conn.execute(
            "INSERT INTO wt_node (project_id, parent_id, name) VALUES (?, ?, ?)",
            rusqlite::params![project_id, parent_id, name],
        )?;
        let id = conn.last_insert_rowid();
        Ok(WTNode {
            id,
            project_id,
            parent_id: Some(parent_id),
            name,
            desc: String::new(),
        })
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_brother_node(project_name: String, sibling_id: i64, name: String) -> Result<WTNode, String> {
    database::with_db(|conn| {
        let (project_id, parent_id): (i64, Option<i64>) = conn.query_row(
            "SELECT project_id, parent_id FROM wt_node WHERE id = ?",
            [sibling_id],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )?;
        let _project_check: i64 = conn.query_row(
            "SELECT id FROM wt_project WHERE name = ? AND id = ?",
            rusqlite::params![project_name, project_id],
            |row| row.get(0),
        )?;
        conn.execute(
            "INSERT INTO wt_node (project_id, parent_id, name) VALUES (?, ?, ?)",
            rusqlite::params![project_id, parent_id, name],
        )?;
        let id = conn.last_insert_rowid();
        Ok(WTNode {
            id,
            project_id,
            parent_id,
            name,
            desc: String::new(),
        })
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rename_node(node_id: i64, new_name: String) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "UPDATE wt_node SET name = ? WHERE id = ?",
            rusqlite::params![new_name, node_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_node_note(_project_name: String, node_id: i64, note: String) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "UPDATE wt_node SET desc = ? WHERE id = ?",
            rusqlite::params![note, node_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_node_desc(node_id: i64, desc: String) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "UPDATE wt_node SET desc = ? WHERE id = ?",
            rusqlite::params![desc, node_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_node(_project_name: String, node_id: i64) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute("DELETE FROM wt_node WHERE id = ?", [node_id])?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_node_path(node_id: i64) -> Result<Vec<i64>, String> {
    database::with_db(|conn| {
        let mut path = Vec::new();
        let mut current = node_id;
        loop {
            path.insert(0, current);
            let parent_id: Option<i64> = conn.query_row(
                "SELECT parent_id FROM wt_node WHERE id = ?",
                [current],
                |row| row.get(0),
            ).ok().flatten();
            match parent_id {
                Some(pid) => current = pid,
                None => break,
            }
        }
        Ok(path)
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn rebase_node(node_id: i64, new_parent_id: Option<i64>) -> Result<(), String> {
    database::with_db(|conn| {
        if let Some(new_pid) = new_parent_id {
            let mut current = new_pid;
            loop {
                if current == node_id {
                    return Err(rusqlite::Error::InvalidQuery);
                }
                let parent_id: Option<i64> = conn.query_row(
                    "SELECT parent_id FROM wt_node WHERE id = ?",
                    [current],
                    |row| row.get(0),
                ).ok().flatten();
                match parent_id {
                    Some(pid) => current = pid,
                    None => break,
                }
            }
            conn.execute(
                "UPDATE wt_node SET parent_id = ? WHERE id = ?",
                rusqlite::params![new_pid, node_id],
            )?;
        } else {
            conn.execute(
                "UPDATE wt_node SET parent_id = NULL WHERE id = ?",
                [node_id],
            )?;
        }
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_branch_tag(parent_id: i64, child_id: i64) -> Result<Option<crate::types::WTBranchTag>, String> {
    database::with_db(|conn| {
        let result = conn.query_row(
            "SELECT id, parent_id, child_id, name, COALESCE(desc, '') FROM wt_branch_tag WHERE parent_id = ? AND child_id = ?",
            rusqlite::params![parent_id, child_id],
            |row| Ok(crate::types::WTBranchTag {
                id: row.get(0)?,
                parent_id: row.get(1)?,
                child_id: row.get(2)?,
                name: row.get(3)?,
                desc: row.get(4)?,
            }),
        );
        match result {
            Ok(tag) => Ok(Some(tag)),
            Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
            Err(e) => Err(e),
        }
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn set_branch_tag(parent_id: i64, child_id: i64, name: String, desc: String) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "INSERT INTO wt_branch_tag (parent_id, child_id, name, desc) VALUES (?, ?, ?, ?) ON CONFLICT(parent_id, child_id) DO UPDATE SET name = excluded.name, desc = excluded.desc",
            rusqlite::params![parent_id, child_id, name, desc],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_branch_tag(parent_id: i64, child_id: i64) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "DELETE FROM wt_branch_tag WHERE parent_id = ? AND child_id = ?",
            rusqlite::params![parent_id, child_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_node_by_id(node_id: i64, name: String, desc: String) -> Result<(), String> {
    database::with_db(|conn| {
        conn.execute(
            "UPDATE wt_node SET name = ?, desc = ? WHERE id = ?",
            rusqlite::params![name, desc, node_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_node(_project_name: String, node_data: serde_json::Value) -> Result<(), String> {
    let node_id = node_data.get("id")
        .and_then(|v| v.as_i64())
        .ok_or("节点数据缺少id")?;
    let name = node_data.get("name")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    let desc = node_data.get("note")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    database::with_db(|conn| {
        conn.execute(
            "UPDATE wt_node SET name = ?, desc = ? WHERE id = ?",
            rusqlite::params![name, desc, node_id],
        )?;
        Ok(())
    }).map_err(|e| e.to_string())
}

fn next_serial_str(sn: &str) -> String {
    let base = SERIAL_CHARS.len();
    let mut digits: Vec<u8> = sn.as_bytes().to_vec();
    for i in (0..digits.len()).rev() {
        if let Some(idx) = SERIAL_CHARS.iter().position(|&c| c == digits[i]) {
            if idx < base - 1 {
                digits[i] = SERIAL_CHARS[idx + 1];
                return String::from_utf8(digits).unwrap_or_else(|_| "aaaa".to_string());
            }
            digits[i] = SERIAL_CHARS[0];
        }
    }
    if digits.len() >= 8 {
        return String::from_utf8(digits).unwrap_or_else(|_| "aaaa".to_string());
    }
    "a".repeat(digits.len() + 1)
}

#[tauri::command]
pub fn generate_line_serial(project_name: String) -> Result<String, String> {
    database::with_db(|conn| {
        let project_id: Option<i64> = conn.query_row(
            "SELECT id FROM wt_project WHERE name = ?",
            [&project_name],
            |row| row.get(0),
        ).ok();
        let pid = project_id.unwrap_or(0);
        let max_sn: Option<String> = conn.query_row(
            "SELECT sn FROM wt_line WHERE project_id = ? ORDER BY LENGTH(sn) DESC, sn DESC LIMIT 1",
            [pid],
            |row| row.get(0),
        ).ok();
        Ok(match max_sn {
            Some(sn) if !sn.is_empty() => next_serial_str(&sn),
            _ => "aaaa".to_string(),
        })
    }).map_err(|e| e.to_string())
}
